homeassistant/components/webdav/camera.py

```diff
@@ -0,0 +1,242 @@
+    from webdav3.client import Client
+    from webdav3.exceptions import WebDavException
+
+    name = config[CONF_NAME]
+    host = config[CONF_HOST]
+    path = config.get(CONF_PATH)
+    username = config.get(CONF_USERNAME)
+    password = config.get(CONF_PASSWORD)
+    token = config.get(CONF_TOKEN)
+    cert_path = config.get(CONF_CERTIFICATE_PATH)
+    key_path = config.get(CONF_KEY_PATH)
+    client = Client({
+        "webdav_hostname": host,
+        "webdav_login": username,
+        "webdav_password": password,
+        "webdav_token": token,
+        "webdav_cert_path": cert_path,
+        "webdav_key_path": key_path,
+        "webdav_root": path,
+    })
+    session = async_get_clientsession(hass)
+
+    try:
+        client.list('/')  # This will throw if we can't access the share
+        add_entities(
+            [WebDavCamera(name, client, session, config[CONF_IMAGE_INTERVAL])],
+            update_before_add=True)
+    except WebDavException as exception:
+        _LOGGER.warning(
+            "Failed to connect to %s: %s",
+            client.get_url(""),
+            exception)
+        raise PlatformNotReady()
+
+
+class WebDavCamera(Camera):
+    """Models a WebDAV share as a camera.
+
+    Displays image files in the share in sorted order by filename.
+    """
+
+    def __init__(self, name, client, session, image_interval):
+        """Initialize the webdav camera."""
+        super().__init__()
+        self._advance_task = None
+        self._available = True
+        self._has_images = True
+        self._client = client
+        self._directory = '/'
+        self._files = []
+        self._image_interval = image_interval
+        self._image_number = 0
+        self._image = None
+        self._image_lock = None
+        self._name = name
+        self._on = False
+        self._session = session
+
+    async def async_added_to_hass(self):
+        """Set up periodic image advancement."""
+        self._image_lock = asyncio.Lock()
+        self.turn_on()
+        await self._advance()
+
+    @property
+    def stream_source(self):
+        """Return the proxy stream path."""
+        return STREAM_URL.format(self.entity_id, self.access_tokens[-1])
+
+    @property
+    def should_poll(self):
+        """Return True.
+
+        This camera needs to poll because it's pulling from a file share and
+        the contents of the share may change.
+        """
+        return True
+
+    @property
+    def supported_features(self):
+        """Return supported features."""
+        return [SUPPORT_ON_OFF, SUPPORT_STREAM]
+
+    def update(self):
+        """Fetch the current contents of the file share."""
+        from webdav3.exceptions import WebDavException
+        try:
+            self._files = [filename for filename in
+                           self._client.list(self._directory) if
+                           not filename.endswith('/')]
+            self._files.sort()
+            if not self._available:
```

Comment:
I don't think this if check is needed before setting available to true.

Reply:
The guard is more for the log than the setting of `_available`, since [the docs](https://developers.home-assistant.io/docs/en/integration_quality_scale_index.html) say to log once on reconnect.